File: packages/xlens-android/src/main/kotlin/com/youthperformance/xlens/network/ConvexClient.kt
// ═══════════════════════════════════════════════════════════════
// CONVEX CLIENT
// HTTP client for Convex backend communication
// Handles mutations, queries, and file uploads
// ═══════════════════════════════════════════════════════════════

package com.youthperformance.xlens.network

import com.youthperformance.xlens.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

/**
 * HTTP client for Convex backend
 * Maps to the jump/*.ts Convex functions
 */
internal class ConvexClient(
    private val baseUrl: String,
    private val authToken: String? = null
) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(60, java.util.concurrent.TimeUnit.SECONDS)
        .writeTimeout(300, java.util.concurrent.TimeUnit.SECONDS) // 5 min for uploads
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    private val queryUrl = "$baseUrl/query"
    private val mutationUrl = "$baseUrl/mutation"
    private val uploadUrl = "$baseUrl/upload"

    // MARK: - Session Management

    @Serializable
    data class CreateSessionResponse(
        val sessionId: String,
        val nonce: String,
        val nonceDisplay: String,
        val expiresAt: Long,
        val expiresInMs: Int
    )

    suspend fun createSession(userId: String, deviceKeyId: String?): CreateSessionResponse {
        val args = buildJsonObject {
            put("userId", userId)
            deviceKeyId?.let { put("deviceKeyId", it) }
        }

        return mutation("jump/sessions:create", args)
    }

    @Serializable
    data class SessionValidationResponse(
        val valid: Boolean,
        val reason: String? = null
    )

    suspend fun validateSession(sessionId: String, nonce: String): SessionValidationResponse {
        val args = buildJsonObject {
            put("sessionId", sessionId)
            put("nonce", nonce)
        }

        return query("jump/sessions:validate", args)
    }

    // MARK: - Jump Submission

    @Serializable
    data class SubmitJumpResponse(
        val jumpId: String,
        val status: String
    )

    suspend fun submitJump(
        userId: String,
        sessionId: String,
        videoStorageId: String,
        sensorStorageId: String,
        proofPayload: ProofPayload,
        gps: GPSLocation?
    ): SubmitJumpResponse {
        val args = buildJsonObject {
            put("userId", userId)
            put("sessionId", sessionId)
            put("videoStorageId", videoStorageId)
            put("sensorStorageId", sensorStorageId)
            put("proofPayload", proofPayload.toJsonObject())
            gps?.let {
                put("gps", buildJsonObject {
                    put("city", it.city)
                    it.state?.let { state -> put("state", state) }
                    put("country", it.country)
                })
            }
        }

        return mutation("jump/jumps:submit", args)
    }

    @Serializable
    data class MarkUploadedResponse(val success: Boolean)

    suspend fun markJumpUploaded(jumpId: String): MarkUploadedResponse {
        val args = buildJsonObject {
            put("jumpId", jumpId)
        }

        return mutation("jump/jumps:markUploaded", args)
    }

    // MARK: - Jump Queries

    suspend fun listJumps(userId: String, limit: Int, excludePractice: Boolean = false): List<Jump> {
        val args = buildJsonObject {
            put("userId", userId)
            put("limit", limit)
            put("excludePractice", excludePractice)
        }

        val response: List<JumpResponse> = query("jump/jumps:listForUser", args)

        return response.map { it.toJump() }
    }

    suspend fun getBestJump(userId: String, minTier: VerificationTier?): Jump? {
        val args = buildJsonObject {
            put("userId", userId)
            minTier?.let { put("minTier", it.name.lowercase()) }
        }

        val response: JumpResponse? = queryNullable("jump/jumps:getBestForUser", args)
        return response?.toJump()
    }

    // MARK: - Daily Cap

    suspend fun checkDailyCap(userId: String): DailyCapStatus {
        val args = buildJsonObject {
            put("userId", userId)
        }

        val response: CapResponse = query("jump/jumpUsers:checkDailyCap", args)

        return DailyCapStatus(
            jumpsUsed = response.jumpsUsed,
            remaining = response.remaining,
            cap = response.cap,
            isOverCap = response.isOverCap
        )
    }

    // MARK: - File Upload

    suspend fun uploadFile(data: ByteArray, contentType: String): String = withContext(Dispatchers.IO) {
        val requestBody = data.toRequestBody(contentType.toMediaType())

        val request = Request.Builder()
            .url(uploadUrl)
            .post(requestBody)
            .apply {
                authToken?.let { addHeader("Authorization", "Bearer $it") }
            }
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            val message = response.body?.string() ?: "Unknown error"
            throw XLensException.UploadFailed("Status ${response.code}: $message")
        }

        val body = response.body?.string()
            ?: throw XLensException.DecodingError("Empty response body")

        val uploadResponse = json.decodeFromString<UploadResponse>(body)
        uploadResponse.storageId
    }

    // MARK: - Generic Query/Mutation

    private suspend inline fun <reified T> query(path: String, args: JsonObject): T {
        return request(queryUrl, path, args)
    }

    private suspend inline fun <reified T> queryNullable(path: String, args: JsonObject): T? {
        return try {
            request(queryUrl, path, args)
        } catch (e: XLensException.DecodingError) {
            null
        }
    }

    private suspend inline fun <reified T> mutation(path: String, args: JsonObject): T {
        return request(mutationUrl, path, args)
    }

    private suspend inline fun <reified T> request(url: String, path: String, args: JsonObject): T = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("path", path)
            put("args", args)
            put("format", "json")
        }

        val requestBody = json.encodeToString(body).toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .addHeader("Content-Type", "application/json")
            .apply {
                authToken?.let { addHeader("Authorization", "Bearer $it") }
            }
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw XLensException.NetworkError(e)
        }

        if (!response.isSuccessful) {
            val message = response.body?.string() ?: "Unknown error"
            throw XLensException.ServerError(response.code, message)
        }

        val responseBody = response.body?.string()
            ?: throw XLensException.DecodingError("Empty response body")

        try {
            // Convex wraps response in { "value": ... }
            val jsonElement = json.parseToJsonElement(responseBody)
            val valueElement = jsonElement.jsonObject["value"]
                ?: throw XLensException.DecodingError("Missing 'value' in response")

            json.decodeFromJsonElement(valueElement)
        } catch (e: Exception) {
            if (e is XLensException) throw e
            throw XLensException.DecodingError(e.message ?: "Unknown decoding error")
        }
    }

    // MARK: - Response Types

    @Serializable
    private data class UploadResponse(val storageId: String)

    @Serializable
    private data class JumpResponse(
        val _id: String,
        val userId: String,
        val sessionId: String,
        val status: String,
        val heightInches: Double? = null,
        val heightCm: Double? = null,
        val flightTimeMs: Int? = null,
        val confidence: Double? = null,
        val verificationTier: String? = null,
        val isPractice: Boolean = false,
        val createdAt: Long = 0
    ) {
        fun toJump() = Jump(
            id = _id,
            status = JumpStatus.valueOf(status.uppercase()),
            sessionId = sessionId,
            heightInches = heightInches,
            heightCm = heightCm,
            flightTimeMs = flightTimeMs,
            verificationTier = verificationTier?.let { VerificationTier.valueOf(it.uppercase()) },
            confidence = confidence
        )
    }

    @Serializable
    private data class CapResponse(
        val jumpsUsed: Int,
        val remaining: Int,
        val cap: Int,
        val isOverCap: Boolean,
        val needsReset: Boolean = false
    )
}

// MARK: - Proof Payload

data class ProofPayload(
    val sessionId: String,
    val nonce: String,
    val capture: CaptureInfo,
    val hashes: Hashes,
    val signature: Signature,
    val gps: GPSCoordinates? = null
) {
    data class CaptureInfo(
        val testType: String = "VERT_JUMP",
        val startedAtMs: Long,
        val endedAtMs: Long,
        val fps: Int,
        val device: DeviceInfo
    )

    data class DeviceInfo(
        val platform: String = "android",
        val model: String,
        val osVersion: String,
        val appVersion: String
    )

    data class Hashes(
        val videoSha256: String,
        val sensorSha256: String,
        val metadataSha256: String
    )

    data class Signature(
        val alg: String = "ES256",
        val keyId: String,
        val sig: String
    )

    data class GPSCoordinates(
        val lat: Double,
        val lng: Double,
        val accuracyM: Double,
        val capturedAtMs: Long
    )

    fun toJsonObject(): JsonObject = buildJsonObject {
        put("sessionId", sessionId)
        put("nonce", nonce)
        put("capture", buildJsonObject {
            put("testType", capture.testType)
            put("startedAtMs", capture.startedAtMs)
            put("endedAtMs", capture.endedAtMs)
            put("fps", capture.fps)
            put("device", buildJsonObject {
                put("platform", capture.device.platform)
                put("model", capture.device.model)
                put("osVersion", capture.device.osVersion)
                put("appVersion", capture.device.appVersion)
            })
        })
        put("hashes", buildJsonObject {
            put("videoSha256", hashes.videoSha256)
            put("sensorSha256", hashes.sensorSha256)
            put("metadataSha256", hashes.metadataSha256)
        })
        put("signature", buildJsonObject {
            put("alg", signature.alg)
            put("keyId", signature.keyId)
            put("sig", signature.sig)
        })
        gps?.let {
            put("gps", buildJsonObject {
                put("lat", it.lat)
                put("lng", it.lng)
                put("accuracyM", it.accuracyM)
                put("capturedAtMs", it.capturedAtMs)
            })
        }
    }
}

File: packages/xlens-ios/Sources/xLENS/Network/ConvexClient.swift
// ═══════════════════════════════════════════════════════════════
// CONVEX CLIENT
// HTTP client for Convex backend communication
// Handles mutations, queries, and file uploads
//
// Swift 2026 Best Practices:
// - Actor isolation for thread safety
// - Sendable nested types
// - Pure async/await (no Combine)
// ═══════════════════════════════════════════════════════════════

import Foundation

/// HTTP client for Convex backend
/// Maps to the jump/*.ts Convex functions
actor ConvexClient {

    // MARK: - Properties

    private let baseURL: URL
    private let authToken: String?
    private let session: URLSession

    // Convex function endpoints
    private var queryURL: URL { baseURL.appendingPathComponent("query") }
    private var mutationURL: URL { baseURL.appendingPathComponent("mutation") }
    private var uploadURL: URL { baseURL.appendingPathComponent("upload") }

    // MARK: - Initialization

    init(baseURL: URL, authToken: String? = nil) {
        self.baseURL = baseURL
        self.authToken = authToken

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 300 // 5 minutes for uploads
        self.session = URLSession(configuration: config)
    }

    // MARK: - Session Management

    struct CreateSessionResponse: Codable, Sendable {
        let sessionId: String
        let nonce: String
        let nonceDisplay: String
        let expiresAt: Int64
        let expiresInMs: Int
    }

    func createSession(userId: String, deviceKeyId: String?) async throws -> CreateSessionResponse {
        var args: [String: Any] = ["userId": userId]
        if let deviceKeyId = deviceKeyId {
            args["deviceKeyId"] = deviceKeyId
        }

        return try await mutation(
            path: "jump/sessions:create",
            args: args
        )
    }

    struct SessionValidationResponse: Codable, Sendable {
        let valid: Bool
        let reason: String?
    }

    func validateSession(sessionId: String, nonce: String) async throws -> SessionValidationResponse {
        return try await query(
            path: "jump/sessions:validate",
            args: [
                "sessionId": sessionId,
                "nonce": nonce
            ]
        )
    }

    // MARK: - Jump Submission

    struct SubmitJumpResponse: Codable, Sendable {
        let jumpId: String
        let status: String
    }

    func submitJump(
        userId: String,
        sessionId: String,
        videoStorageId: String,
        sensorStorageId: String,
        proofPayload: ProofPayload,
        gps: GPSLocation?
    ) async throws -> SubmitJumpResponse {
        var args: [String: Any] = [
            "userId": userId,
            "sessionId": sessionId,
            "videoStorageId": videoStorageId,
            "sensorStorageId": sensorStorageId,
            "proofPayload": proofPayload.toDictionary()
        ]

        if let gps = gps {
            var gpsDict: [String: Any] = [
                "city": gps.city,
                "country": gps.country
            ]
            if let state = gps.state {
                gpsDict["state"] = state
            }
            args["gps"] = gpsDict
        }

        return try await mutation(
            path: "jump/jumps:submit",
            args: args
        )
    }

    struct MarkUploadedResponse: Codable, Sendable {
        let success: Bool
    }

    func markJumpUploaded(jumpId: String) async throws -> MarkUploadedResponse {
        return try await mutation(
            path: "jump/jumps:markUploaded",
            args: ["jumpId": jumpId]
        )
    }

    // MARK: - Jump Queries

    func listJumps(userId: String, limit: Int, excludePractice: Bool = false) async throws -> [Jump] {
        struct JumpResponse: Codable, Sendable {
            let _id: String
            let userId: String
            let sessionId: String
            let status: String
            let heightInches: Double?
            let heightCm: Double?
            let flightTimeMs: Int?
            let confidence: Double?
            let verificationTier: String?
            let isPractice: Bool
            let createdAt: Int64
        }

        let response: [JumpResponse] = try await query(
            path: "jump/jumps:listForUser",
            args: [
                "userId": userId,
                "limit": limit,
                "excludePractice": excludePractice
            ]
        )

        return response.map { j in
            Jump(
                id: j._id,
                status: JumpStatus(rawValue: j.status) ?? .processing,
                sessionId: j.sessionId,
                heightInches: j.heightInches,
                heightCm: j.heightCm,
                flightTimeMs: j.flightTimeMs,
                verificationTier: j.verificationTier.flatMap { VerificationTier(rawValue: $0) },
                confidence: j.confidence
            )
        }
    }

    func getBestJump(userId: String, minTier: VerificationTier?) async throws -> Jump? {
        struct JumpResponse: Codable, Sendable {
            let _id: String
            let userId: String
            let sessionId: String
            let status: String
            let heightInches: Double?
            let heightCm: Double?
            let flightTimeMs: Int?
            let confidence: Double?
            let verificationTier: String?
        }

        var args: [String: Any] = ["userId": userId]
        if let minTier = minTier {
            args["minTier"] = minTier.rawValue
        }

        let response: JumpResponse? = try await query(
            path: "jump/jumps:getBestForUser",
            args: args
        )

        guard let j = response else { return nil }

        return Jump(
            id: j._id,
            status: JumpStatus(rawValue: j.status) ?? .processing,
            sessionId: j.sessionId,
            heightInches: j.heightInches,
            heightCm: j.heightCm,
            flightTimeMs: j.flightTimeMs,
            verificationTier: j.verificationTier.flatMap { VerificationTier(rawValue: $0) },
            confidence: j.confidence
        )
    }

    // MARK: - Daily Cap

    func checkDailyCap(userId: String) async throws -> DailyCapStatus {
        struct CapResponse: Codable, Sendable {
            let jumpsUsed: Int
            let remaining: Int
            let cap: Int
            let isOverCap: Bool
            let needsReset: Bool
        }

        let response: CapResponse = try await query(
            path: "jump/jumpUsers:checkDailyCap",
            args: ["userId": userId]
        )

        return DailyCapStatus(
            jumpsUsed: response.jumpsUsed,
            remaining: response.remaining,
            cap: response.cap,
            isOverCap: response.isOverCap
        )
    }

    // MARK: - File Upload

    func uploadFile(data: Data, contentType: String) async throws -> String {
        var request = URLRequest(url: uploadURL)
        request.httpMethod = "POST"
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")

        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (responseData, response) = try await session.upload(for: request, from: data)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw XLensError.networkError("Invalid response type")
        }

        guard httpResponse.statusCode == 200 else {
            let message = String(data: responseData, encoding: .utf8) ?? "Unknown error"
            throw XLensError.uploadFailed("Status \(httpResponse.statusCode): \(message)")
        }

        struct UploadResponse: Codable, Sendable {
            let storageId: String
        }

        let uploadResponse = try JSONDecoder().decode(UploadResponse.self, from: responseData)
        return uploadResponse.storageId
    }

    // MARK: - Generic Query/Mutation

    private func query<T: Decodable>(path: String, args: [String: Any]) async throws -> T {
        return try await request(url: queryURL, path: path, args: args)
    }

    private func mutation<T: Decodable>(path: String, args: [String: Any]) async throws -> T {
        return try await request(url: mutationURL, path: path, args: args)
    }

    private func request<T: Decodable>(url: URL, path: String, args: [String: Any]) async throws -> T {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let body: [String: Any] = [
            "path": path,
            "args": args,
            "format": "json"
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw XLensError.networkError("Invalid response type")
        }

        guard httpResponse.statusCode == 200 else {
            let message = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw XLensError.serverError(httpResponse.statusCode, message)
        }

        // Convex wraps response in { "value": ... }
        struct ConvexResponse<V: Decodable>: Decodable {
            let value: V
        }

        do {
            let convexResponse = try JSONDecoder().decode(ConvexResponse<T>.self, from: data)
            return convexResponse.value
        } catch {
            throw XLensError.decodingError(error.localizedDescription)
        }
    }
}

// MARK: - Proof Payload

struct ProofPayload: Sendable {
    let sessionId: String
    let nonce: String
    let capture: CaptureInfo
    let hashes: Hashes
    let signature: Signature
    let gps: GPSCoordinates?

    struct CaptureInfo: Sendable {
        let testType: String = "VERT_JUMP"
        let startedAtMs: Int64
        let endedAtMs: Int64
        let fps: Int
        let device: DeviceInfo
    }

    struct DeviceInfo: Sendable {
        let platform: String = "ios"
        let model: String
        let osVersion: String
        let appVersion: String
    }

    struct Hashes: Sendable {
        let videoSha256: String
        let sensorSha256: String
        let metadataSha256: String
    }

    struct Signature: Sendable {
        let alg: String = "ES256"
        let keyId: String
        let sig: String
    }

    struct GPSCoordinates: Sendable {
        let lat: Double
        let lng: Double
        let accuracyM: Double
        let capturedAtMs: Int64
    }

    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "sessionId": sessionId,
            "nonce": nonce,
            "capture": [
                "testType": capture.testType,
                "startedAtMs": capture.startedAtMs,
                "endedAtMs": capture.endedAtMs,
                "fps": capture.fps,
                "device": [
                    "platform": capture.device.platform,
                    "model": capture.device.model,
                    "osVersion": capture.device.osVersion,
                    "appVersion": capture.device.appVersion
                ]
            ],
            "hashes": [
                "videoSha256": hashes.videoSha256,
                "sensorSha256": hashes.sensorSha256,
                "metadataSha256": hashes.metadataSha256
            ],
            "signature": [
                "alg": signature.alg,
                "keyId": signature.keyId,
                "sig": signature.sig
            ]
        ]

        if let gps = gps {
            dict["gps"] = [
                "lat": gps.lat,
                "lng": gps.lng,
                "accuracyM": gps.accuracyM,
                "capturedAtMs": gps.capturedAtMs
            ]
        }

        return dict
    }
}
